src/rust-crypto/dehydration.ts

```diff
@@ -0,0 +1,262 @@
+*/
+
+import * as RustSdkCryptoJs from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { RustCrypto } from "./rust-crypto";
+import { OutgoingRequestProcessor } from "./OutgoingRequestProcessor";
+import { encodeUri } from "../utils";
+import { ClientPrefix, IHttpOpts, MatrixError, MatrixHttpApi, Method } from "../http-api";
+import { IToDeviceEvent } from "../sync-accumulator";
+import { ServerSideSecretStorage } from "../secret-storage";
+import { crypto } from "../crypto/crypto";
+import { decodeBase64, encodeUnpaddedBase64 } from "../base64";
+import { Logger } from "../logger";
+import { UnstableValue } from "../NamespacedValue";
+
+// schema for the API response bodies
+interface IDehydratedDeviceResp {
+    device_id: string;
+    device_data: {
+        algorithm: string;
+    };
+}
+interface IDehydratedDeviceEventsResp {
+    events: IToDeviceEvent[];
+    next_batch: string;
+}
+
+export const UnstablePrefix = ClientPrefix.Unstable + "/org.matrix.msc3814.v1";
+const SECRET_STORAGE_NAME = new UnstableValue("m.dehydrated_device", "org.matrix.msc3814");
+
+/**
+ * Manages dehydrated devices
+ */
+export class RustDehydrationManager {
+    private key?: Uint8Array;
+    private intervalId?: ReturnType<typeof setInterval>;
+    private timeoutId?: ReturnType<typeof setTimeout>;
+
+    public constructor(
+        private readonly logger: Logger,
+        private readonly rustCrypto: RustCrypto,
+        private readonly olmMachine: RustSdkCryptoJs.OlmMachine,
+        private readonly http: MatrixHttpApi<IHttpOpts & { onlyData: true }>,
+        private readonly outgoingRequestProcessor: OutgoingRequestProcessor,
+        private readonly secretStorage: ServerSideSecretStorage,
+    ) {}
+
+    /** Return whether the server supports dehydrated devices
+     */
+    public async isSupported(): Promise<boolean> {
+        // call the endpoint to get a dehydrated device.  If it returns an
+        // M_UNRECOGNIZED error, then dehydration is unsupported.  If it returns
+        // a successful response, or an M_NOT_FOUND, then dehydration is supported.
+        // Any other exceptions are passed through.
+        try {
+            await this.http.authedRequest<IDehydratedDeviceResp>(
+                Method.Get,
+                "/dehydrated_device",
+                undefined,
+                undefined,
+                {
+                    prefix: UnstablePrefix,
+                },
+            );
+        } catch (error) {
+            const err = error as MatrixError;
+            if (err.errcode === "M_UNRECOGNIZED") {
+                return false;
+            } else if (err.errcode === "M_NOT_FOUND") {
+                return true;
+            }
+            throw error;
+        }
+        return true;
+    }
+
+    /** Return whether the dehydration key is stored in Secret Storage
+     */
+    public async isKeyStored(): Promise<boolean> {
+        return Boolean(await this.secretStorage.isStored(SECRET_STORAGE_NAME.name));
```

Comment:
I think you might want to look for `SECRET_STORAGE_NAME.altName` as well as `name` here, for forward compatibility, so this code works when the stable name is in use. For example, we often use `NamespacedValue.matches` to check both. What we're doing here seems conceptually similar to `NamespacedValue.findIn` although I don't think we can actually use it here.